torch/distributed/_tensor/experimental/attention.py

```diff
@@ -0,0 +1,467 @@
+
+            out_grad_queries.append(grad_query)
+            out_grad_keys.append(grad_key)
+            out_grad_values.append(grad_value)
+
+    # stack and sum to avoid accumulation errors
+    out_grad_query = torch.stack(out_grad_queries).sum(dim=0)
+    out_grad_key = torch.stack(out_grad_keys).sum(dim=0)
+    out_grad_value = torch.stack(out_grad_values).sum(dim=0)
+
+    return out_grad_query, out_grad_key, out_grad_value
+
+
+customized_ops = {
+    aten._scaled_dot_product_flash_attention.default: sdpa_handler,
+    aten._scaled_dot_product_flash_attention_backward.default: sdpa_backward_handler,
+}
+
+
+@contextlib.contextmanager
+def attention_context_parallel() -> Generator[None, None, None]:
+    """
+    This enables attention context parallel optimizations. Currently only supports ring attention for SDPA flash attention.
+    """
+    DTensor._op_dispatcher._custom_op_handlers.update(customized_ops)
+
+    yield
+
+    for custom_op in customized_ops:
+        DTensor._op_dispatcher._custom_op_handlers.pop(custom_op)
+
+
+class ContextParallel(ParallelStyle):
+    """
+    Applies context parallel optimizations to the attention layer. This will
+    work for nn.MultiHeadedAttention.
+
+    Currently only supports ring attention for SDPA flash attention.
+
+    See attention_context_parallel for more details.
+    """
+
+    def _apply(self, module: nn.Module, device_mesh: DeviceMesh) -> nn.Module:
+        if not isinstance(module, nn.MultiheadAttention):
+            raise ValueError(f"{type(module)} is not supported by {type(self)} yet.")
+
+        if not isinstance(device_mesh, DeviceMesh):
+            raise ValueError(
+                f"{type(device_mesh)} is not supported by {type(self)} yet."
+            )
+
+        if not device_mesh.ndim == 1:
+            raise ValueError
+
+        return distribute_module(
+            module,
+            device_mesh,
+            input_fn=self._input_fn,  # type: ignore[arg-type]
+            output_fn=self._output_fn,  # type: ignore[arg-type]
+        )
+
+    @staticmethod
+    def _input_fn(
+        module: nn.Module,
+        inputs: Tuple[torch.Tensor, torch.Tensor, torch.Tensor],
+        device_mesh: DeviceMesh,
+    ) -> Tuple[DTensor, DTensor, DTensor]:
+        # TODO(d4l3k); this should be Shard(2), need to fix Linear layer rules
+        query_placement = [Replicate()]
+        kv_placement = [Replicate()]
+
+        # convert inputs to DTensor
+        query, key, value = inputs
+        if not isinstance(query, DTensor):
+            query = DTensor.from_local(
+                query, device_mesh, query_placement, run_check=False
+            )
+        if not isinstance(key, DTensor):
+            key = DTensor.from_local(key, device_mesh, kv_placement, run_check=False)
+        if not isinstance(value, DTensor):
+            value = DTensor.from_local(
+                value, device_mesh, kv_placement, run_check=False
+            )
```

Comment:
I recalled that we would need to call context manager.__enter__ and __exit__ for input_fn and output_fn so that user don't need to manually activate the context manager, wondering why we removed it somehow?

Reply:
ah yeah, that was in the design doc but I didn't actually implement it -- will add it

Reply:
done -- there's some interesting behavior here with this. I had to use register_hook on the tensors to enable the backwards ops correctly